soc: sam0: Kill off dts_fixup.h extras

Sizurka opened the pull request:
We can't kill the file entirely, as some samples/tests still use a few defines from it, but this brings it down the the "normal" level that other SoCs are at.

The first commit just moves everything to automatically generated DT defines which had no reason to be in dts_fixup.h.  The second part removes the SERCOM parts, which involves adding DTS aliases for the devices.  This allows the existing mapping for SERCOM to DTS definition to be preserved, though through a slightly modified name.

This was tested on some of the basic API tests (e.g. `tests/drivers/spi/spi_loopback` and other samples), but it shouldn't actually be changing any functionality.

benpicco commented on soc/arm/atmel_sam0/samd21/soc.h:
```diff
@@ -68,4 +68,21 @@
+#define _SOC_ATMEL_SAM0_SERCOM_CHECK(dt_index, type, sercom_id) \
+	(IS_ENABLED(DT_ATMEL_SAM0_ ## type ## _ ## dt_index) && \
+	(DT_ATMEL_SAM0_ ## type ## _ ## dt_index ## _SERCOM_ID) == sercom_id)
+#define SOC_ATMEL_SAM0_SERCOM_IS_ENABLED(type, id) (	\
```
I think this should be moved to `atmel_sam0/common/soc_sercom.h`

Sizurka replied:
I've changed it to use direct DTS aliases, which removes the need for those macros entirely.  However, this means supporting SAME54 depends on #14897 to support the extra interrupts.

Sizurka committed:
soc: sam0: Remove SERCOM defines from dts_fixup.h

Move SERCOM peripherals to use the raw defines generated from DTS
parsing.  This adds aliases to the DTS so that the SERCOM number
can still be used for clocking and pinmux.

Signed-off-by: Derek Hageman <hageman@inthat.cloud>

benpicco reviewed (approved):
It took me a bit to understand the significance of those aliases, but this indeed simplifies things considerably!

benpicco commented on drivers/watchdog/wdt_sam0.c:
```diff
@@ -143,23 +141,23 @@ static int wdt_sam0_install_timeout(struct device *dev,
 				per += 1;
 			}
-			WDT_REGS->EWCTRL.bit.EWOFFSET = per - 1;
+			WDT->EWCTRL.bit.EWOFFSET = per - 1;
```
This should be `WDT->EWCTRL.bit.EWOFFSET = per - 1U;` now after 24d71431e9c9da493ce00eb284ac9b0b2cb1bdca

Sizurka replied:
ACK and rebased.

MaureenHelm reviewed (changes requested):
Please split the first commit into separate commits for each peripheral type

Sizurka committed:
drivers: pinmux: sam0: Remove defines from dts_fixup.h

Move SAM0 pinmux to use the raw defines generated from the DTS
parsing.

Signed-off-by: Derek Hageman <hageman@inthat.cloud>

Sizurka committed:
drivers: gpio: sam0: Remove defines from dts_fixup.h

Move SAM0 GPIO to use the raw defines generated from the DTS
parsing.

Signed-off-by: Derek Hageman <hageman@inthat.cloud>

Sizurka committed:
drivers: flash: sam0: Remove defines from dts_fixup.h

Move SAM0 flash to use the raw defines generated from the DTS
parsing.

Signed-off-by: Derek Hageman <hageman@inthat.cloud>

Sizurka committed:
drivers: watchdog: sam0: Remove defines from dts_fixup.h

Move SAM0 GPIO to use the raw defines generated from the DTS
parsing.

Signed-off-by: Derek Hageman <hageman@inthat.cloud>

Sizurka committed:
drivers: timer: sam0: Remove RTC defines from dts_fixup.h

Move SAM0 flash to use the raw defines generated from the DTS
parsing.

Signed-off-by: Derek Hageman <hageman@inthat.cloud>

Sizurka committed:
drivers: usb: sam0: Remove defines from dts_fixup.h

Move SAM0 flash to use the raw defines generated from the DTS
parsing.

Signed-off-by: Derek Hageman <hageman@inthat.cloud>

Sizurka commented:
Done, the non-SERCOM peripherals are split out now.

MaureenHelm commented on dts/arm/atmel/samd.dtsi:
```diff
@@ -33,6 +33,11 @@
 
+	aliases {
+		pinmux-a = &pinmux_a;
+		pinmux-b = &pinmux_b;
```
Why did you add aliases for pinmux and sercom, but not gpio? The generated dts instance indices aren't guaranteed to match the static hardware indices. Consider the case that gpio port a is disabled in dts, then gpio port b will be the first enabled instance in dts and get index 0.

Sizurka replied:
I only added them because they where the only ones used.  That is the current boards only reference the pinmux rather than the GPIO devices.  The GPIO driver itself is independent of the indexing.  I have no problem adding them for GPIO however.

That is, since the GPIO drivers are self contained, the indexing doesn't matter (i.e. it doesn't matter if index 0 refers to PORT A or PORT B, since it has the correct address from the DTS).  Basically, nothing in the GPIO instance has to reference anything looked up in absolute indices (e.g. clocking) so it's not required.

MaureenHelm commented on drivers/watchdog/wdt_sam0.c:
```diff
@@ -12,8 +12,6 @@
 #include <logging/log.h>
 LOG_MODULE_REGISTER(wdt_sam0);
 
-#define WDT_REGS ((Wdt *)DT_WDT_SAM0_BASE_ADDRESS)
```
Please keep using the base address from dts rather than from ASF. We are doing this by convention for all the vendors

Sizurka replied:
ACK, changed.

MaureenHelm commented on drivers/timer/sam0_rtc_timer.c:
```diff
@@ -20,7 +20,7 @@
 
 /* RTC registers. */
-#define RTC0 ((RtcMode0 *) DT_RTC_SAM0_BASE_ADDRESS)
+#define RTC0 (&(RTC->MODE0))
```
Please keep using the base address from dts rather than from ASF. We are doing this by convention for all the vendors

Sizurka replied:
ACK, changed.

MaureenHelm commented on dts/bindings/rtc/atmel,sam0-rtc.yaml:
```diff
@@ -22,10 +22,4 @@ properties:
       generation: define
       category: required
-
-    clock-generator:
```
Please keep this

Sizurka replied:
Ok, I will put it back in, but right now (pending something like #15585) actually setting it would break the driver: it currently hard codes the clock frequency as the one from GCLK 0.

MaureenHelm commented on drivers/usb/device/usb_dc_sam0.c:
```diff
@@ -21,8 +21,7 @@ LOG_MODULE_REGISTER(usb_dc_sam0);
 
 #define USB_SAM0_IN_EP 0x80
 
-#define REGS ((Usb *)DT_USB_DC_SAM0_BASE_ADDRESS)
```
Please keep using the base address from dts rather than from ASF. We are doing this by convention for all the vendors

Sizurka replied:
ACK, changed.

MaureenHelm reviewed (changes requested):
> Done, the non-SERCOM peripherals are split out now.

Thank you! This makes it easier to review

Sizurka commented:
Shippable failure appears to be because #15023 was merged and #14861 isn't yet merged (I can't add labels, or I'd have added a blocked label to #15023, #14867 should likely have it added).  So I don't think it's related.
